### ISSUE TITLE
OSError: Search pattern never detected in send_command_expect: (again)

### ISSUE THREAD

**Issue by jeffsilverm** [ktbyers/netmiko#513]
I hate to bring this up again, because you've kinda beaten to death in Issue 292, Issue 337, Issue 310, Issue 189,....  I am running version 1.4.1 on Python 3.5.2 on Ubuntu 16.04.2 with the 4.4.0-59-generic kernel.

I have a command that executes very slowly, copying a file using FTP.  Typical copy times are on the order of 30 minutes (1,800 seconds) when I do this interactively.

```
miko_obj = ConnectHandler(device_type='juniper_junos', ip=dut, username=admin_user, password=admin_password)
 result_sc = miko_obj.send_command ( "file copy /var/log/messages ftp://jeffsilverman:XXXX@10.3.151.17//tmp/messages_tmp", delay_factor=20, max_loops = 500)
```

I have been futzing with delay_factor and max_loops, and using more and more extreme values.  messages is 314 MBytes, and the command that I have given above will raise OSError.  By way of contrast, the following call will run for tens of minutes:
```
>>> result_sc = miko_obj.send_command ( "file copy /var/log/lmpd ftp://roger:Eric@10.3.151.17//tmp/lmpd_342", delay_factor=20, max_loops = 500)
```

By way of contrast,
```
>>> result_sc = miko_obj.send_command_expect ( "file copy /var/log/lmpd ftp://roger:Eric@10.3.151.17//tmp/lmpd_342", delay_factor=20, max_loops = 500, expect_string="> ")
>>> print(result_sc)
ftp://roger:Eric@10.3.151.17//tmp/messages_tmp100% of  314 MB  160 kBps 00m00s

v-jefsi@str-ptx1k-02>

v-jefsi@str-ptx1k-02> file copy /var/log/messages ftp://roger:Eric@10.3.151.17//tmp/messages_tmp

>>>
>>> result_sc = miko_obj.send_command_expect ( "file copy /var/log/lmpd ftp://roger:Eric@10.3.151.17//tmp/lmpd_342", expect_string="> ")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/v-jefsi/.local/lib/python3.5/site-packages/netmiko/base_connection.py", line 823, in send_command_expect
    return self.send_command(*args, **kwargs)
  File "/home/v-jefsi/.local/lib/python3.5/site-packages/netmiko/base_connection.py", line 815, in send_command
    search_pattern))
OSError: Search pattern never detected in send_command_expect: >
>>>

```
The contents of ```result_sc``` suggests to me that the file copy actually did finish, but for some reason, paramiko didn't detect it.  When paramiko raises OSError, what happens to the command executing on the router?  Does it keep running, and eventually returns something even though the client has already bailed out on it?

I have been inspecting the source code - you're a better pythonista than I am - and I do not see an explanation of delay_factor.  I see your code modifying it for all sorts of reasons.  max_loops I think I understand.  Changing max_loops counts the number of times the code checks for the prompt string before raising OSError.  I don't actually see where OSError is raised, I assume that paramiko is doing that.

I've seen this issue discussed before, but I don't have a solution that works for me.  I'm sorry.

Jeff

**Comment by jeffsilverm** [comment]
I am going to try read_command and write_command.  Give me about 4 hours, please.

**Comment by ktbyers** [comment]
@jeffsilverm FYI, I think the command above didn't actually complete in the time allocated:

```
>>> print(result_sc)
ftp://roger:Eric@10.3.151.17//tmp/messages_tmp100% of  314 MB  160 kBps 00m00s

v-jefsi@str-ptx1k-02>

v-jefsi@str-ptx1k-02> file copy /var/log/messages ftp://roger:Eric@10.3.151.17//tmp/messages_tmp

```

**Comment by ktbyers** [comment]
Does this work. Note, I changed the expect string below to be a raw string r">" and eliminated the sapce below:

```
result_sc = miko_obj.send_command_expect ( "file copy /var/log/lmpd ftp://roger:Eric@10.3.151.17//tmp/lmpd_342", expect_string=r">", max_loops=1000, delay_factor=20)
```

**Comment by ktbyers** [comment]
The delay_factor basically multiplies all the delays. So setting to 2 will double the delays; setting to 4 will quadruple the delays.

By default send_command/send_command_expect will wait for about 90 seconds before giving up (delay_factor = 1).

max_loops also can be used to adjust how long it waits.

On my ToDo list is to unify the timeouts into a more logical timeout, but it will probably be Sept/Oct before I get to this.

**Comment by ktbyers** [comment]
I am pretty sure it is failing here:

https://github.com/ktbyers/netmiko/blob/96b133a84d4f4ee1847351d24d90de7cf60e1921/netmiko/base_connection.py#L810

On quick glance the OSError vs IOError is probably due to some Python changes.

**Comment by jeffsilverm** [comment]
Keith,

That worked, except I had to change the max_loops to 2000.

I am contemplating setting the max_loops to a number, 2000, and putting in an exception handler that increases max_loops by a certain amount (TBD) every time the OSError exception is raised.  Yes, that is a bit of a kludge, and I am open to a better suggestion.  Since I am using this in an automated test system which is running more or less in batch mode, if it takes a few minutes longer, than that's acceptable.

Changing the expect_string to a raw string did not make a difference.  Removing the space *did* make a difference, although the space is there.  I have no explanation, but that's okay.

I think you are right about the OSError change.  Could that be a change from paramiko?  If your users are running with different versions of paramiko or if the exception is dependent on the underlying python or worse, the underlying operating system, then that will be a minor pain in the butt but not an insolvable problem.

I really am in love with netmiko.  I would like to write more tutorial documentation if you are open to that sort of thing.

**Comment by ktbyers** [comment]
@jeffsilverm Okay, nice.

What I would like to do is make a more general timeout where you set the `timeout` on connect and then Netmiko would wait that long for a command to complete. So if your command took 10 minutes, you could just set the timeout to wait long enough.

...but probably not going to get to this until September/October.

Yes you could have some sort of try/except that increased `max_loops` or you could just set it to a large value. It will finish quicker if it can (so it is just a matter of how long you want your program to wait before generating an error).

I think the `OSError` is a Python change (on quick glance). If OSError is going to be a new pattern from Python, then I should probably raise a different exception as `OSError` is misleading.

Yes, feel free to write more tutorial documentation.

I also have done some work on documentation of methods (but much more is needed). So this is another place documentation work is needed:

http://netmiko.readthedocs.io/en/latest/

Basically, you add docstrings formatted in a certain way and that gets automatically converted into readable documentation.

**Comment by tee-lake** [comment]
@ktbyers @jeffsilverm 
As a clarification, will "NetMikoTimeoutException" be the right way to handle the timeout issue for a slow device such that max_loop parameter can be automatically adjusted. Can you please give a sample code snippet that illustrates the automatic adjustment?

Many Thanks

**Comment by ktbyers** [comment]
The exception would possibly be IOError. There are many contexts where you can get an error due to not giving more time so you need to probably delimit it (i.e. are you talking about `send_command` in which case, you will probably get IOError).

I probably wouldn't automatically adjust, I would probably just call send_command using higher `delay_factor` and `max_loops` arguments.

For example:
```
results = net_connect.send_command("cmd", max_loops=1000, delay_factor=20)
```

You could do an initial call and then catch the error and try it again, but I probably wouldn't do this in practice:

```
try:
    results = net_connect.send_command("cmd")
except IOError:
    net_connect.clear_buffer()
    results = net_connect.send_command("cmd", max_loops=1000, delay_factor=20)
```

**Comment by RakeshPhython** [comment]
HI Krick,
I have written below code but it is not running  well, I am getting 


Trying to connect10.10.50.5
**************************************

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-12-db31fa4d5af9> in <module>()
     42 for ip in Two_Switch_IP:
     43     Two_River_Switch=switch(ip)
---> 44     switch_conn=Two_River_Switch.connect_to_device()
     45     show_cdp_nei_ip=Two_River_Switch.get_first_cdp_nei_ip()
     46     show_vstack_cmd=Two_River_Switch.get_show_opt_vstack()

<ipython-input-12-db31fa4d5af9> in connect_to_device(self)
     22             print("**************************************")
     23             try:
---> 24                 self.conn = ConnectHandler(**self.TwoRiver_Connect_Dict)
     25             except AttributeError:
     26                 print("Unable to connect to {}".format(ip))

~\Anaconda3\lib\site-packages\netmiko\ssh_dispatcher.py in ConnectHandler(*args, **kwargs)
    176                          'currently supported platforms are: {}'.format(platforms_str))
    177     ConnectionClass = ssh_dispatcher(kwargs['device_type'])
--> 178     return ConnectionClass(*args, **kwargs)
    179 
    180 

~\Anaconda3\lib\site-packages\netmiko\base_connection.py in __init__(self, ip, host, username, password, secret, port, device_type, verbose, global_delay_factor, use_keys, key_file, allow_agent, ssh_strict, system_host_keys, alt_host_keys, alt_key_file, ssh_config_file, timeout, session_timeout, blocking_timeout, keepalive, default_enter, response_return, serial_settings)
    206             self._modify_connection_params()
    207             self.establish_connection()
--> 208             self.session_preparation()
    209 
    210     def __enter__(self):

~\Anaconda3\lib\site-packages\netmiko\cisco\cisco_ios.py in session_preparation(self)
     15         """Prepare the session after the connection has been established."""
     16         self._test_channel_read(pattern=r'[>#]')
---> 17         self.set_base_prompt()
     18         self.disable_paging()
     19         self.set_terminal_width(command='terminal width 511')

~\Anaconda3\lib\site-packages\netmiko\base_connection.py in set_base_prompt(self, pri_prompt_terminator, alt_prompt_terminator, delay_factor)
    852         :type delay_factor: int
    853         """
--> 854         prompt = self.find_prompt(delay_factor=delay_factor)
    855         if not prompt[-1] in (pri_prompt_terminator, alt_prompt_terminator):
    856             raise ValueError("Router prompt not found: {0}".format(repr(prompt)))

~\Anaconda3\lib\site-packages\netmiko\base_connection.py in find_prompt(self, delay_factor)
    893         prompt = prompt.strip()
    894         if not prompt:
--> 895             raise ValueError("Unable to find prompt: {}".format(prompt))
    896         time.sleep(delay_factor * .1)
    897         self.clear_buffer()

ValueError: Unable to find prompt: 


or some time it connect to device successfully but do not provide proper out put 

Trying to connect10.10.50.5
**************************************
Able to connect successfully10.10.50.5
**************************show cdp neighbhor of switch***************************************


show cdp neighbhor of switch10.10.50.5is
['x#terminal']
************************** show vstack output****************************************

show vstack output['-DSW-4500x#', 'TR-RM-DC-DSW-4500x#', 'TR-RM-DC-DSW-4500x#', 'TR-RM-DC-DSW-4500x#']is
**************************Disconnetng from switch****************************************

Disconnetng from switch10.10.50.5

 Good bye 10.10.50.5

Trying to connect10.10.50.110
**************************************
Trying to connect10.10.50.5
**************************************
Able to connect successfully10.10.50.5
**************************show cdp neighbhor of switch***************************************


show cdp neighbhor of switch10.10.50.5is
['x#terminal']
************************** show vstack output****************************************

show vstack output['-DSW-4500x#', 'TR-RM-DC-DSW-4500x#', 'TR-RM-DC-DSW-4500x#', 'TR-RM-DC-DSW-4500x#']is
**************************Disconnetng from switch****************************************

Disconnetng from switch10.10.50.5

 Good bye 10.10.50.5

Trying to connect10.10.50.110
**************************************
Able to connect successfully10.10.50.110

Socket exception: An established connection was aborted by the software in your host machine (10053)

---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
<ipython-input-15-db31fa4d5af9> in <module>()
     43     Two_River_Switch=switch(ip)
     44     switch_conn=Two_River_Switch.connect_to_device()
---> 45     show_cdp_nei_ip=Two_River_Switch.get_first_cdp_nei_ip()
     46     show_vstack_cmd=Two_River_Switch.get_show_opt_vstack()
     47     print("**************************show cdp neighbhor of switch***************************************")

<ipython-input-15-db31fa4d5af9> in get_first_cdp_nei_ip(self)
     31     def get_first_cdp_nei_ip(self):
     32         cdp_nei_cmd = 'sh cdp neighbors'
---> 33         cdp_nei_det_op=self.conn.send_command(cdp_nei_cmd)
     34         return cdp_nei_det_op
     35     def get_show_opt_vstack(self):

~\Anaconda3\lib\site-packages\netmiko\base_connection.py in send_command(self, command_string, expect_string, delay_factor, max_loops, auto_find_prompt, strip_prompt, strip_command, normalize, use_textfsm)
   1007             if auto_find_prompt:
   1008                 try:
-> 1009                     prompt = self.find_prompt(delay_factor=delay_factor)
   1010                 except ValueError:
   1011                     prompt = self.base_prompt

~\Anaconda3\lib\site-packages\netmiko\base_connection.py in find_prompt(self, delay_factor)
    884                     prompt = self.strip_ansi_escape_codes(prompt).strip()
    885             else:
--> 886                 self.write_channel(self.RETURN)
    887                 time.sleep(delay_factor * .1)
    888             count += 1

~\Anaconda3\lib\site-packages\netmiko\base_connection.py in write_channel(self, out_data)
    288         self._lock_netmiko_session()
    289         try:
--> 290             self._write_channel(out_data)
    291         finally:
    292             # Always unlock the SSH channel, even on exception.

~\Anaconda3\lib\site-packages\netmiko\base_connection.py in _write_channel(self, out_data)
    266         """
    267         if self.protocol == 'ssh':
--> 268             self.remote_conn.sendall(write_bytes(out_data))
    269         elif self.protocol == 'telnet':
    270             self.remote_conn.write(write_bytes(out_data))

~\Anaconda3\lib\site-packages\paramiko\channel.py in sendall(self, s)
    829         """
    830         while s:
--> 831             sent = self.send(s)
    832             s = s[sent:]
    833         return None

~\Anaconda3\lib\site-packages\paramiko\channel.py in send(self, s)
    783         m.add_byte(cMSG_CHANNEL_DATA)
    784         m.add_int(self.remote_chanid)
--> 785         return self._send(s, m)
    786 
    787     def send_stderr(self, s):

~\Anaconda3\lib\site-packages\paramiko\channel.py in _send(self, s, m)
   1168                 # this doesn't seem useful, but it is the documented behavior
   1169                 # of Socket
-> 1170                 raise socket.error('Socket is closed')
   1171             size = self._wait_for_send_window(size)
   1172             if size == 0:

OSError: Socket is closed






Code ---- To connect to Network device and run command

import xlrd
import netmiko
from netmiko import ConnectHandler
import time
from netmiko.ssh_exception import AuthenticationException,NetMikoTimeoutException
wb=xlrd.open_workbook("Two_River.xlsx")
sheet_name=wb.sheet_by_index(0)
Two_Switch_IP=sheet_name.col_values(1)
Two_Switch_Hostname=sheet_name.col_values(0)

class switch():                            
    def __init__(self,ip,username='test',password='test',device_type='cisco_ios'):
        self.ip=ip
        self.TwoRiver_Connect_Dict={'username':username,
                                    'password':password,
                                    'device_type':device_type,
                                     'ip':self.ip
                                    }
      
    def connect_to_device(self):
            print("Trying to connect{}".format(ip))
            print("**************************************")
            try:
                self.conn = ConnectHandler(**self.TwoRiver_Connect_Dict)
            except AttributeError:
                print("Unable to connect to {}".format(ip))
            except NetMikoTimeoutException:
                print("Unable to connect to {}".format(ip))
            else:
                print("Able to connect successfully{}".format(ip))
    def get_first_cdp_nei_ip(self):
        cdp_nei_cmd = 'sh cdp neighbors'
        cdp_nei_det_op=self.conn.send_command(cdp_nei_cmd)
        return cdp_nei_det_op
    def get_show_opt_vstack(self):
        cdp_vstack_cmd = 'show vstack config'
        vstack_cmd_op=self.conn.send_command(cdp_vstack_cmd)
        return vstack_cmd_op 
    def switch_disconnect(self):
        self.conn.clear_buffer()
        self.conn.disconnect() 
for ip in Two_Switch_IP:
    Two_River_Switch=switch(ip)
    switch_conn=Two_River_Switch.connect_to_device()
    show_cdp_nei_ip=Two_River_Switch.get_first_cdp_nei_ip()
    show_vstack_cmd=Two_River_Switch.get_show_opt_vstack()
    print("**************************show cdp neighbhor of switch***************************************")
    print("\n")
    print("show cdp neighbhor of switch{}is".format(ip))
    print(show_cdp_nei_ip.splitlines())
    print("************************** show vstack output****************************************\n")
    print("show vstack output{}is".format(show_vstack_cmd.splitlines()))
    print("**************************Disconnetng from switch****************************************\n")
    print("Disconnetng from switch{}\n".format(ip))
    Two_River_Switch.switch_disconnect()
    print(" Good bye {}\n".format(ip))
